### ISSUE TITLE
load starlark from target output

### ISSUE THREAD

**Issue by izissise** [facebook/buck2#710]
Hello,

I'm currently migrating a repository from Please Build to Buck2, and I've encountered an issue with `load` statements.

In Please, there is a feature called `subinclude` (see [Please Builtins](https://please.build/lexicon.html#please-builtins)), which inserts the output of the target passed as a parameter to `subinclude` at the position it is called. This allows for some neat tricks, such as generating genrule statements programmatically.

I haven't found an equivalent feature in Buck2. Does a similar feature exist? If not, how challenging would it be to add this functionality?

Thanks,
Hugues

**Comment by toothbrush7777777** [comment]
Not as far as I know. However, you can feed a `genrule`'s input from another `genrule`. You produce an output file from the first `genrule` and then feed that as the input to the next `genrule`.

I am unaware of any facility to dynamically produce Starlark rules, though, since Buck2 explicitly separates the analysis and compilation phases.
Maybe you could also look at [dynamic outputs/dependencies](https://buck2.build/docs/rule_authors/dynamic_dependencies/)?

**Comment by stepancheg** [comment]
@izissise buck2 does not build targets or execute commands during target graph construction.

There are reasons for that, for example, target determinator, which should be fast and reliable.

When build definitions need to be generated, we generate them outside of buck2, and internally commit them to the repository.

This feature could be potentially implemented, but we don’t have plans for it.

**Comment by stepancheg** [comment]
@toothbrush7777777 dynamic output allows accessing command output during build time (delayed/dynamic) analysis, but does not allow creation of new targets (or even subtargets).

It can be used to sort dependencies between files within target, for example, but cannot even create undeclared dependencies.

**Comment by toothbrush7777777** [comment]
> @toothbrush7777777 dynamic output allows accessing command output during build time (delayed/dynamic) analysis, but does not allow creation of new targets (or even subtargets).
> 
> It can be used to sort dependencies between files within target, for example, but cannot even create undeclared dependencies.

Yes, but we've found that is often enough when combined with macros. That's why I suggested it. Obviously the relevant targets/rules would need to be adapted to fit better with Buck2's methodology, but I was trying to suggest a starting point if rules needed to be created from more dynamic rules.

**Comment by izissise** [comment]
Do you think I can a make PR were I would try a proof of concept for this?

**Comment by stepancheg** [comment]
@izissise I'd suggest instead:

- first, describe your use case, and why it cannot be solved with other tools (like for example, by generating `BUCK` files before execution of `buck2` or by committing generated files to the repository).
- then maybe write a proposal

Proof of concept is a lot of work. Month of work easily: it change buck2 model quite significantly. And there's more than one way to solve the problem you solve. `subinclude` as "Please" does it may be not the best way to do this feature.

Also maintaining this feature we will unlikely to use internally, is significant work. So I'm skeptical about it, although there's some demand for it (for example, for [reindeer](https://github.com/facebook/buck2/blob/7bf69c4a8aeb9c69e2f3098438fc156c436d1cb1/.github/workflows/build-and-test.yml#L51)).

**Comment by izissise** [comment]
Generating BUCK files before execution or committing generated files to the repository isn't ideal. Using a single tool maintains consistency and prevents stale files that need regeneration. I have not found any way to load a file in the parser from the target output. I'm interested in any other way to do this or to achieve similar results.

use case:

- In Please, I override genrule to produce a Starlark file that, when subincluded, loads a variable with all the parameters given to the genrule. This allows retrieval of meta information on a target from elsewhere in the tree. Example:

```

# boot/platform/aarch64-qemu/BUCK
meta_genrule(
    name='linux',
    meta={
        "arch": 'aarch64',
        "linux_release": "6.6",
    },
    srcs={
        "UKI": [uki],
        "QEMU": [qemu],
    },
)

# BUCK

def meta_get(pkg, target):
    subinclude(f'{pkg}:{target}_+meta')
    return meta_dict # from subinclude

meta = meta_get(pkg='boot/platform/aarch64-qemu', target='linux')
linux_release = meta['linux_release']

```


- Another use case is dynamically transforming Makefiles or Cargo.toml (or other build systems) into Buck2 rules:
```
transform_rule = transfrom_makefile('Makefile')
subinclude(transform_rule)
reexport = genrule(name='makefile', srcs=[makefile])
```
I believe this is similar to the reeinder feature you mentionned.


I'm sure there are more, adding dynamism will allow to convert any inputs (given the correct tools) into a buck2 genrule graph.
Also I'm willing to get more involve in buck2 development, it might be a way in.

Thanks for the discussion,
Hugues

**Comment by izissise** [comment]
@stepancheg I'm re-visiting this issue, I still think having this dynamic behavior would allow to make buck2 compatible with simpler build tools like make while still letting buck2 handling cache.

If this is something that people want, maybe there is a grant system so someone could work on this?

**Comment by alexlian** [comment]
As @stepancheg noted, this is very much against the direction we are headed internally, and the flexibility that comes with this comes at the tradeoff of scaling and performance.

At this point, will close won't fix, as this doesn't seem like a direction we would like to compromise on at least for now.
